src/rust/engine/options/src/lib.rs

```diff
@@ -297,86 +304,129 @@ impl OptionParser {
+        default: Vec<T>,
+        getter: fn(&Rc<dyn OptionsSource>, &OptionId) -> Result<Option<Vec<ListEdit<T>>>, String>,
+    ) -> Result<Vec<T>, String> {
+        self.parse_list(id, default, getter, |list, remove| {
+            let to_remove = remove.iter().collect::<HashSet<_>>();
+            list.into_iter()
+                .filter(|item| !to_remove.contains(item))
+                .collect()
         })
     }
 
+    pub fn parse_bool_list(&self, id: &OptionId, default: &[bool]) -> Result<Vec<bool>, String> {
+        self.parse_list_hashable(id, default.to_vec(), |source, id| source.get_bool_list(id))
+    }
+
+    pub fn parse_int_list(&self, id: &OptionId, default: &[i64]) -> Result<Vec<i64>, String> {
+        self.parse_list_hashable(id, default.to_vec(), |source, id| source.get_int_list(id))
+    }
+
+    // Floats are not Eq or Hash, so we fall back to the brute-force O(N*M) lookups.
+    pub fn parse_float_list(&self, id: &OptionId, default: &[f64]) -> Result<Vec<f64>, String> {
+        self.parse_list(
+            id,
+            default.to_vec(),
+            |source, id| source.get_float_list(id),
+            |list, to_remove| {
+                list.into_iter()
+                    .filter(|item| !to_remove.contains(item))
+                    .collect()
+            },
+        )
+    }
+
     pub fn parse_string_list(
         &self,
         id: &OptionId,
         default: &[&str],
     ) -> Result<Vec<String>, String> {
-        let mut list_edits = vec![];
+        self.parse_list_hashable::<String>(
+            id,
+            default.iter().map(|s| s.to_string()).collect(),
+            |source, id| source.get_string_list(id),
+        )
+    }
+
+    pub fn parse_dict(
+        &self,
+        id: &OptionId,
+        default: HashMap<String, Val>,
+    ) -> Result<HashMap<String, Val>, String> {
+        let mut dict_edits: Vec<DictEdit> = vec![];
         for (_, source) in self.sources.iter() {
-            if let Some(edits) = source.get_string_list(id)? {
-                list_edits.extend(edits);
+            if let Some(edit) = source.get_dict(id)? {
+                dict_edits.push(edit);
             }
         }
-        let mut string_list = default.iter().map(|s| s.to_string()).collect::<Vec<_>>();
-        for list_edit in list_edits {
-            match list_edit.action {
-                ListEditAction::Replace => string_list = list_edit.items,
-                ListEditAction::Add => string_list.extend(list_edit.items),
-                ListEditAction::Remove => {
-                    let to_remove = list_edit.items.iter().collect::<HashSet<_>>();
-                    string_list = string_list
-                        .iter()
-                        .filter(|item| !to_remove.contains(item))
-                        .map(|s| s.to_owned())
-                        .collect::<Vec<String>>();
-                }
+        let mut dict = default;
+        for dict_edit in dict_edits {
```

Comment:
Minor stylistic comment is that I think some extra whitespace is always welcome! I'd have added at least a blank line after the first loop, and probably one after this loop. Makes the logical blocks clearer.

(And workflow wise - it's much nicer to navigate airy code when one uses whitespace-based navigation, i.e., emacs `next-paragraph` etc.)